run.go

```diff
@@ -139,38 +135,102 @@ var commandInspect = cli.Command{
-	switch {
-	case strings.Contains(filename, "://"):
-		return TypeURL
-	case strings.HasSuffix(filename, ".js"):
-		return TypeJS
-	default:
-		return ""
+func looksLikeURL(str []byte) bool {
+	// Try to parse string as URL
+	u, err := url.Parse(strings.TrimSpace(string(str)))
+	if err != nil || u.Scheme == "" {
+		return false
 	}
+	return true
 }
 
-func makeRunner(filename, t string) (lib.Runner, error) {
-	if t == TypeAuto {
-		t = guessType(filename)
+func getSrcData(arg, t string) (*lib.SourceData, error) {
+	srcdata := []byte("")
+	srctype := t
+	filename := arg
+	var err error
+	const cmdline = "[cmdline]"
+	// special case name “-“ will always cause src data to be read from file STDIN
+	if arg == "-" {
+		srcdata, err = ioutil.ReadAll(os.Stdin)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		// Deduce how to get src data
+		switch srctype {
+		case lib.TypeAuto:
+			if looksLikeURL([]byte(arg)) { // always try to parse as URL string first
+				srcdata = []byte(arg)
+				srctype = lib.TypeURL
+				filename = cmdline
+			} else {
+				// Otherwise, check if it is a file name and we can load the file
+				srcdata, err = ioutil.ReadFile(arg)
+				if err != nil { // if we fail to open file, we assume the arg is JS code
+					srcdata = []byte(arg)
+					srctype = lib.TypeJS
+					filename = cmdline
+				}
+			}
+		case lib.TypeURL:
+			// We try to use lib.TypeURL args as URLs directly first
+			if looksLikeURL([]byte(arg)) {
+				srcdata = []byte(arg)
+				filename = cmdline
+			} else { // if that didn’t work, we try to load a file with URLs
+				srcdata, err = ioutil.ReadFile(arg)
+				if err != nil {
+					return nil, err
+				}
+			}
+		case lib.TypeJS:
+			// lib.TypeJS args we try to use as file names first
+			srcdata, err = ioutil.ReadFile(arg)
+			if err != nil { // and if that didn’t work, we assume the arg itself is JS code
+				srcdata = []byte(arg)
+				filename = cmdline
+			}
+		}
 	}
+	// Now we should have some src data and in most cases a type also. If we
+	// don’t have a type it means we read from STDIN or from a file and the user
+	// specified type == lib.TypeAuto. This means we need to try and auto-detect type
+	if srctype == lib.TypeAuto {
+		if looksLikeURL(srcdata) {
+			srctype = lib.TypeURL
+		} else {
+			srctype = lib.TypeJS
+		}
+	}
+	src := &lib.SourceData{
+		SrcData:  srcdata,
+		Filename: filename,
+		SrcType:  srctype,
+	}
+	return src, nil
+}
 
-	switch t {
+func makeRunner(srcdata *lib.SourceData) (lib.Runner, error) {
+	switch srcdata.SrcType {
 	case "":
-		return nil, errors.New("Unable to infer type from argument; specify with -t/--type")
-	case TypeURL:
-		r, err := simple.New(filename)
+		return nil, errors.New("Invalid type specified, see --help")
+	case lib.TypeURL:
+		u, err := url.Parse(strings.TrimSpace(string(srcdata.SrcData)))
+		if err != nil || u.Scheme == "" {
+			return nil, errors.New("Failed to parse URL")
+		}
+		r, err := simple.New(u)
```

Comment:
The whole thought was to put as much user input verification stuff as possible in one single place (run.go), rather than both there and inside the runners. It felt like a win also to not have to use url.Parse() in two places (run.go and simple/runner.go). The JS runner doesn't work the same way - there it is the runtime function Load() that I changed to instead take the already loaded (supposedly) JS code instead of a file name. I.e. the simple runner and the JS runtime both get more refined/verified data as arguments, than they used to, which should be somewhat less likely to cause them trouble. But if you think this is a bad idea I'll move it back into the runner.